homeassistant/components/media_player/songpal.py

```diff
@@ -0,0 +1,250 @@
+
+                await device.async_set_sound_setting(params[PARAM_NAME],
+                                                     params[PARAM_VALUE])
+
+    hass.services.async_register(
+        DOMAIN, SET_SOUND_SETTING, async_service_handler,
+        schema=SET_SOUND_SCHEMA)
+
+
+class SongpalDevice(MediaPlayerDevice):
+    """Class representing a Songpal device."""
+
+    def __init__(self, name, endpoint):
+        """Init."""
+        import songpal
+        self._name = name
+        self.endpoint = endpoint
+        self.dev = songpal.Protocol(self.endpoint)
+        self._sysinfo = None
+
+        self._state = False
+        self._available = False
+        self._initialized = False
+
+        self._volume_control = None
+        self._volume_min = 0
+        self._volume_max = 1
+        self._volume = 0
+        self._is_muted = False
+
+        self._sources = []
+
+    async def initialize(self):
+        """Initialize the device."""
+        await self.dev.get_supported_methods()
+        self._sysinfo = await self.dev.get_system_info()
+
+    @property
+    def name(self):
+        """Return name of the device."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Return an unique ID."""
+        return self._sysinfo.macAddr
+
+    @property
+    def available(self):
+        """Return availability of the device."""
+        return self._available
+
+    async def async_set_sound_setting(self, name, value):
+        """Change a setting on the device."""
+        await self.dev.set_sound_settings(name, value)
+
+    async def async_update(self):
+        """Fetch updates from the device."""
+        from songpal import SongpalException
+        try:
+            volumes = await self.dev.get_volume_information()
+            if not volumes:
+                _LOGGER.error("Got no volume controls, bailing out")
+                return
+            if len(volumes) > 1:
+                _LOGGER.warning("Got %s volume controls, using the first one",
+                                volumes)
+                return
+
+            volume = volumes.pop()
+            _LOGGER.debug("Current volume: %s", volume)
+            self._volume_max = volume.maxVolume
+            self._volume_min = volume.minVolume
+            self._volume = volume.volume
+            self._volume_control = volume
+            self._is_muted = self._volume_control.is_muted
+
+            status = await self.dev.get_power()
+            self._state = status.status
+            _LOGGER.debug("Got state: %s", status)
+
+            inputs = await self.dev.get_inputs()
+            _LOGGER.debug("Got ins: %s", inputs)
+            self._sources = inputs
+
+            self._available = True
+        except SongpalException as ex:
+            # if we were available, print out the exception
+            if self._available:
+                _LOGGER.error("Got an exception %s", ex, exc_info=True)
```

Comment:
Use `_LOGGER.exception`, it will log the exception.

Does this also mean that whenever the device goes offline, we log a whole stack trace? That seems undesirable.

Reply:
That it will do indeed. I thought it'd be a good idea to have something visible in case some of the above calls fail, but I can remove it and just print out the exception.